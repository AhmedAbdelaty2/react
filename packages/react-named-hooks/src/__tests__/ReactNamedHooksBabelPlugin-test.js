/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

const babel = require('@babel/core');
const {wrap} = require('jest-snapshot-serializer-raw');
const namedHooks = require('react-named-hooks');

function transform(input, options = {}) {
  return wrap(
    babel.transform(input, {
      babelrc: false,
      configFile: false,
      plugins: [
        '@babel/syntax-jsx',
        [
          namedHooks,
          {
            skipEnvCheck: true,
            // To simplify debugging tests:
            emitFullSignatures: true,
            ...options.freshOptions,
          },
        ],
        ...(options.plugins || []),
      ],
    }).code,
  );
}

describe('ReactNamedHooksBabelPlugin', () => {
  it('injects first array item for useState, useReducer', () => {
    expect(
      transform(`
        function Foo() {
          const [count, setCount] = useState(0);

          const [data, dispatch] = useReducer(reducer, initialData);
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });

  it('do not injects when lowercased first array item is a default hook name for useState, useReducer', () => {
    expect(
      transform(`
        function Foo() {
          const [State, setCount] = useState(0);

          const [reduceR, dispatch] = useReducer(reducer, initialData);
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });

  it('inejcts identifier for useRef, useCallback, useMemo', () => {
    expect(
      transform(`
        function Foo() {
          const spanRef = useRef(null);

          const memoizedSetClick = useCallback(() => setCount(count + 1), [count]);

          const memoizedCountMultiplied = useMemo(() => count * 2, [count]);
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });

  it('do not injects when lowercased identifier is a default hook name for useRef, useCallback, useMemo', () => {
    expect(
      transform(`
        function Foo() {
          const Ref = useRef(null);

          const CallBack = useCallback(() => setCount(count + 1), [count]);

          const memo = useMemo(() => count * 2, [count]);
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });

  it('inejcts first argument for useContext', () => {
    expect(
      transform(`
        function Foo() {
          const ctxValue = useContext(StringContext);
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });

  it('do not injects when lowercased first argument is a default hook name for useContext', () => {
    expect(
      transform(`
        function Foo() {
          const ctxValue = useContext(Context);
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });

  it('injects for default hooks within nested custom hook', () => {
    expect(
      transform(`
        function useNestedInnerHook() {
          const [nestedState] = useState(123);
          return nestedState;
        }
        function useNestedOuterHook() {
          return useNestedInnerHook();
        }
        function Foo() {
          useNestedOuterHook();
          return <div></div>;
        }
    `),
    ).toMatchSnapshot();
  });
});
