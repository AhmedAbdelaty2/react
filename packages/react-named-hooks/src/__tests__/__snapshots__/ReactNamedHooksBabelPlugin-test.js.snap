// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactNamedHooksBabelPlugin injects first array item for useState, useReducer 1`] = `
function Foo() {
  const [count, setCount] = useState(0);
  useDebugName("count");
  const [data, dispatch] = useReducer(reducer, initialData);
  useDebugName("data");
  return <div></div>;
}
`;

exports[`ReactNamedHooksBabelPlugin do not injects when lowercased first array item is a default hook name for useState, useReducer 1`] = `
function Foo() {
  const [State, setCount] = useState(0);
  const [reduceR, dispatch] = useReducer(reducer, initialData);
  return <div></div>;
}
`;

exports[`ReactNamedHooksBabelPlugin inejcts identifier for useRef, useCallback, useMemo 1`] = `
function Foo() {
  const spanRef = useRef(null);
  useDebugName("spanRef");
  const memoizedSetClick = useCallback(() => setCount(count + 1), [count]);
  useDebugName("memoizedSetClick");
  const memoizedCountMultiplied = useMemo(() => count * 2, [count]);
  useDebugName("memoizedCountMultiplied");
  return <div></div>;
}
`;

exports[`ReactNamedHooksBabelPlugin do not injects when lowercased identifier is a default hook name for useRef, useCallback, useMemo 1`] = `
function Foo() {
  const Ref = useRef(null);
  const CallBack = useCallback(() => setCount(count + 1), [count]);
  const memo = useMemo(() => count * 2, [count]);
  return <div></div>;
}
`;

exports[`ReactNamedHooksBabelPlugin inejcts first argument for useContext 1`] = `
function Foo() {
  const ctxValue = useContext(StringContext);
  useDebugName("StringContext");
  return <div></div>;
}
`;

exports[`ReactNamedHooksBabelPlugin do not injects when lowercased first argument is a default hook name for useContext 1`] = `
function Foo() {
  const ctxValue = useContext(Context);
  return <div></div>;
}
`;

exports[`ReactNamedHooksBabelPlugin injects for default hooks within nested custom hook 1`] = `
function useNestedInnerHook() {
  const [nestedState] = useState(123);
  useDebugName("nestedState");
  return nestedState;
}

function useNestedOuterHook() {
  return useNestedInnerHook();
}

function Foo() {
  useNestedOuterHook();
  return <div></div>;
}
`;
